### PR TITLE
rfc: limit handling in distsql

### PR DESCRIPTION
**Pull request by RaduBerinde**
Initial draft.

### CODE REVIEW

**Comment by knz** [comment]
Thanks for writing this.

One aspect I am missing from this discussion is to reveal opportunities due to non-deterministic ordering when ORDER BY is *not present*. A clause like `select * from foo where x = 10 limit 10` (assuming no index on x) can also be satisfied by launching a distributed read / filter, and abort the distributed computation as soon as 10 results have been received by the gateway. I think there are use cases where this would be beneficial.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldLumB7fksO7o-5oXe:-KldLumCn4cQSdv8ARPo:b-ass13n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L37)):*
> ```Markdown
>      `scanNode`.
> 
>      *Note: The DistSQL sorter is aware of partial orders but the classic SQL
> ```

s/classic/local

---

*[docs/RFCS/distsql_limit.md, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldMEGQYLU3L4UzgaA7:-KldMEGQYLU3L4UzgaA8:b-m5en6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L64)):*
> ```Markdown
>    instances corresponding to a planNode) output rows to streams, in parallel
>    with the consumers of those streams. There is no sequencing step where a
>    processor can stop and wait and see if the consumer needs more rows. In
> ```

Are you saying there is now flow control? I think this is a good occasion to explain how flow control is organized, and discuss whether we need more flow control.

---

*[docs/RFCS/distsql_limit.md, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldMVaVvvbwdcmPsIU_:-KldMVaVvvbwdcmPsIUa:b-ass13n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L85)):*
> ```Markdown
> 
> The goal for this RFC is to find ways to work around the hurdles above to
> achieve performance comparable with Classic SQL in all limit cases.
> ```

s/classic/local

---

*[docs/RFCS/distsql_limit.md, line 108 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldMlG6ofCwIY_B8K87:-KldMlG6ofCwIY_B8K88:b-ydpthk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L108)):*
> ```Markdown
> ### Short-term proposal ###
> 
> The simple proposal is to not distribute TableReaders for limited scans.
> ```

Probably this should only be done when the LIMIT value is small, e.g. less than 10000 rows. Beyond that, distribution might still be helpful.

---

*[docs/RFCS/distsql_limit.md, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldMrxqv_RcqSvqUoRt:-KldMrxqv_RcqSvqUoRu:b-ass13n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L115)):*
> ```Markdown
> 
> We believe that this solution is good for many common cases. It is at least as
> good as classic SQL for queries that only read rows from a single range. It also
> ```

s/classic/local

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-KldN89AKW1V5fdfVhsB:b36oa0n)*

**Comment by RaduBerinde** [comment]
Thanks!

In many of these cases, we still don't want to distribute. If there's a good chance of a single node returning 10 rows, then we would be wasting a lot of work if we went to more nodes. It might help latency of a single query but it will be bad for overall throughput. It's tricky to decide.

There are though some other optimizations we can do in this case:
 - first read through all the ranges held by this node, then fallback to reading other ranges
 - choose the node which holds the most ranges

I will think about it some more and write down something.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldMEGQYLU3L4UzgaA7:-KleHY4cbkP1mTHSogEt:b-626fnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L64)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Are you saying there is now flow control? I think this is a good occasion to explain how flow control is organized, and discuss whether we need more flow control.
</blockquote></details>

There is flow control provided by buffered channels (local streams) and gRPC (remote streams). But the gRPC window sizes are large (to achieve throughput on high latency links).

---

*[docs/RFCS/distsql_limit.md, line 108 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KldMlG6ofCwIY_B8K87:-KleHtN9SAoFnd9ULfzg:b-p7nmip) ([raw file](https://github.com/cockroachdb/cockroach/blob/0909a93cc0f1fd961c72d8f14ba28cf5dbd65f40/docs/RFCS/distsql_limit.md#L108)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Probably this should only be done when the LIMIT value is small, e.g. less than 10000 rows. Beyond that, distribution might still be helpful.
</blockquote></details>

Yes. I remember writing a cutoff like that for index selection.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-KleI1biMqikWjLeOSXq:b-i156vf)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 102 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klj4RCGukAeAEVGAetU:-Klj4RCGukAeAEVGAetV:b7glm5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ad5ee74071a8e805430bfb1305e10dcebbf0a57/docs/RFCS/distsql_limit.md#L102)):*
> ```Markdown
> ## Planning ##
> 
> We have two proposals here: on is simple and should be implemented in the
> ```

`s/on is/one is/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287)*

**Comment by RaduBerinde** [comment]
Updated.

**Comment by arjunravinarayan** [comment]
I think this RFC should be a little more explicit that the goal is to reduce wasteful computation, but care should be taken to explicitly highlight this tradeoff, and perhaps have knobs to shut it down for users who don't mind throwing CPU at the problem for maximum performance.

---

Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 157 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KlsiLEmt5XCJPoah9Vx:-KlsiLEmt5XCJPoah9Vy:b-2g40gw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L157)):*
> ```Markdown
> 
> This solution comes with a planning question: on which node do we put this
> planning synchronizer? A reasonable first implementation is to always use the
> ```

I think *planning synchronizers* need to be generalized a little. Eventually, with materialized views, we will have every processor preceded with a planning synchronizer on each of its inputs, so that all processors can reconstruct their entire input subtrees upon a dead input processor. This is complicated by scenarios where we have a full bisection of N^2 streams (e.g. a hashjoin spread across 5 nodes, followed by a filter stage). If one of the hashjoiners (w.l.o.g. say on node 3) dies, who is responsible for bringing it back up? We have 5 downstream processors (likely only 4, since if node 3 died, then the downstream filternode on node3 also died).

I think fault tolerance is more complicated. It probably deserves an RFC of its own.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-Klsk6jPinecSPBCXPxC:b-a5y61x)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 65 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt69eX1c_yArxIWKFG:-Klss3UCpksAvaqdzLTZ:b-dj7bpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L65)):*
> ```Markdown
>    with the consumers of those streams. We have flow control on streams -
>    provided by buffered channels on local streams, and by gRPC on remote
>    streams; the gRPC window sizes are quite large, necessary for high
> ```

Should we do some sort of slow-start for these window sizes? gRPC doesn't currently support varying these window sizes over time but we might be able to build that in if it would make things better at other levels. (I'm thinking we probably don't want to rely too much on network/rpc-level flow control here, but it's a possibility if it would be helpful). 

---

*[docs/RFCS/distsql_limit.md, line 121 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt7qYR3bNgycppNlLx:-Klt7qYR3bNgycppNlLy:b-nqlvn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L121)):*
> ```Markdown
> has the advantage that we only need to resolve one range during planning. The
> main problem is that in bad cases the table data will be "routed" through that
> first node; this can be very bad if that first node happens to be in a far away
> ```

It could also be bad if that range becomes a hotspot. Local execution would spread the load across all the gateway nodes in the system, while this plan might concentrate load on certain ranges (for example, the first range of a table). Some of the work will inevitably fall on that range sine that's where some of the data lives, but it might be better not to concentrate it further.

---

*[docs/RFCS/distsql_limit.md, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt8R1pqrXyw4lL-AeT:-Klt8R1pqrXyw4lL-AeU:b-9sijb9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L132)):*
> ```Markdown
> In cases like `SELECT * FROM t LIMIT 10`, we can return any 10 rows; so we are
> free to read ranges in any order. An optimization to the proposed solution is to
> place the single TableReader on the node that has most ranges for the table, and
> ```

s/ranges/range leases/

This seems to concentrate the load unnecessarily. I think here we'd ideally want to target the least-loaded nearby node that has the lease for at least one range in the table. But in any case the exact choices made here don't matter too much because we have a lot of flexibility to change int he future. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-Klt9IXROwc5PzZt9Ovp:bb6ywmx)*

**Comment by andreimatei** [comment]
:lgtm:

Thanks for writing this

---

Review status: 0 of 1 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 74 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KlsvgbMhMY0lkURF5sI:-KlsvgbMhMY0lkURF5sJ:b-jbbkyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L74)):*
> ```Markdown
>    if we don't need to.
>  - there are multiple instances of TableReaders for a given table, each working
>    on the ranges are held by that node. Without a special way to plan limits, we
> ```

s/are held/held.

Also there's a stray new line at the end of the prev paragraph.

---

*[docs/RFCS/distsql_limit.md, line 84 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt69d-jezAOoMJwHSN:-Klt69d-jezAOoMJwHSO:bb81yoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L84)):*
> ```Markdown
> useful classes of statements that have limits but where we can't put an (even
> optimistic) limit on scanNodes; for example
> `SELECT * FROM t ORDER BY v LIMIT 10`, where we have to read the entire table in
> ```

nit: you might want to clarify that these examples still have a limit, but at a higher level

---

*[docs/RFCS/distsql_limit.md, line 159 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt8KSWGBCSdJ_iiAEM:-Klt8KSWGBCSdJ_iiAEN:bzha4t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L159)):*
> ```Markdown
> planning synchronizer? A reasonable first implementation is to always use the
> gateway node. More advanced decisions are possible here though (e.g. if the
> table is confined to a datacenter, the synchronizer should be in that
> ```

s/datacenter/zone (as that's the crdb word that captures locality)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-KltA4AUtP4VrLQskp3F:br0oywv)*

**Comment by asubiotto** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 65 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt69eX1c_yArxIWKFG:-KltBYlWC0oSIFii1tDJ:b-v3p4ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L65)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should we do some sort of slow-start for these window sizes? gRPC doesn't currently support varying these window sizes over time but we might be able to build that in if it would make things better at other levels. (I'm thinking we probably don't want to rely too much on network/rpc-level flow control here, but it's a possibility if it would be helpful). 
</blockquote></details>

In https://github.com/cockroachdb/cockroach/issues/14948, @petermattis mentioned that dynamic window sizing is on grpc-go's roadmap.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-KltGd_CahhJLoblGT08:bnfp4nl)*

File: docs/RFCS/distsql_limit.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+### Local SQL ###
+
+There are two types of limits that can apply to a planNode:
+ - a *hard limit* is one where we know we only need **X** rows from a node.
+   Example: `SELECT * FROM t LIMIT 10`
+ - a *soft limit* is one where we have only an optimistic estimation for how
+   many rows we need. Notable cases which lead to soft limits:
+
+   * limits with filters. `SELCT * FROM t WHERE color = 'red' LIMIT 10`. Here
```
Comment (irfansharif) [inline docs/RFCS/distsql_limit.md]: `s/SELCT/SELECT`

**Comment by RaduBerinde** [comment]
Review status: 0 of 1 files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 121 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt7qYR3bNgycppNlLx:-Km1hR2mk2e-uZdgLRVR:b-jevs6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L121)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It could also be bad if that range becomes a hotspot. Local execution would spread the load across all the gateway nodes in the system, while this plan might concentrate load on certain ranges (for example, the first range of a table). Some of the work will inevitably fall on that range sine that's where some of the data lives, but it might be better not to concentrate it further.
</blockquote></details>

Thanks, added a note. Local execution would spread part of the load, but the first range is still a hotspot (reading the first range from KV is roughly equivalent with a tablereader on that node). Ideally we would want our planner to be load-aware and don't put other processors on that node.

---

*[docs/RFCS/distsql_limit.md, line 132 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt8R1pqrXyw4lL-AeT:-Km1j4qj6F2S8rM8IM1p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L132)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/ranges/range leases/

This seems to concentrate the load unnecessarily. I think here we'd ideally want to target the least-loaded nearby node that has the lease for at least one range in the table. But in any case the exact choices made here don't matter too much because we have a lot of flexibility to change int he future. 
</blockquote></details>

Done.

---

*[docs/RFCS/distsql_limit.md, line 157 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KlsiLEmt5XCJPoah9Vx:-Km1jDaxZfXlr9bTIyG5:byluylt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L157)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I think *planning synchronizers* need to be generalized a little. Eventually, with materialized views, we will have every processor preceded with a planning synchronizer on each of its inputs, so that all processors can reconstruct their entire input subtrees upon a dead input processor. This is complicated by scenarios where we have a full bisection of N^2 streams (e.g. a hashjoin spread across 5 nodes, followed by a filter stage). If one of the hashjoiners (w.l.o.g. say on node 3) dies, who is responsible for bringing it back up? We have 5 downstream processors (likely only 4, since if node 3 died, then the downstream filternode on node3 also died).

I think fault tolerance is more complicated. It probably deserves an RFC of its own.
</blockquote></details>

Yes, this is something we will want to extend. The scope needed for this RFC is fairly limited (only supports planning table readers) so the details are more straightforward. There are a lot of complications for a more general primitive (as you point out); I agree this would need a separate RFC.

---

*[docs/RFCS/distsql_limit.md, line 159 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-Klt8KSWGBCSdJ_iiAEM:-Km1j71zBYvJoqhxSqYK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L159)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/datacenter/zone (as that's the crdb word that captures locality)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287)*

**Comment by arjunravinarayan** [comment]
:LGTM: Would you like to move this to the final comment period and make the forum post?

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/distsql_limit.md, line 157 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-KlsiLEmt5XCJPoah9Vx:-Km8IaSpw67wVrmCEFEy:b-ba7jcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccf0445566e039a21ae92dccd431774242c8e82c/docs/RFCS/distsql_limit.md#L157)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, this is something we will want to extend. The scope needed for this RFC is fairly limited (only supports planning table readers) so the details are more straightforward. There are a lot of complications for a more general primitive (as you point out); I agree this would need a separate RFC.
</blockquote></details>

Sounds good to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16287#-:-Km8HKT4Mx3jE2kKCPCW:bx4ixgy)*

**Comment by RaduBerinde** [comment]
Done, entering final comment period.

**Pull request merged**